Add image/source to operating system (#458)
openshift/openshift-origin-design: designs/virtualization/4.6/Add-os-image-to-common-template/add-os-image-to-common-template 4.6.md

---
parent: Virtualization
version: 4.6
---
# Add an operating system image to a common template

## About

We strive to provide an off-the-shelf OS image with every common template.
This an ongoing effort.
Hopefully this will provide the user with a much easier experience to creating a virtual machine.

> This OS image takes the shape of a disk image(data-volume) and can be attached to the OS template (one of the parameters within the common templates).

Reviewer: one thing we'll need to think about is when common templates ship with an image already. This feature maybe become less useful or maybe they can add multiple sources for a single OS depending on the boot source type. 
It could also be for cases where we don't ship a a boot source with a common template they can see that they can add the source themselves (Likely for Windows VMs)

Author: I agree, and I believe this design is an "in-between" solution that will soon not be the main path.

If we go with adding an "Add source" from the VM templates list view or details page, we still need to cover what happens when the user clicks on "Create a VM (from a template without image)".
One alternative to this design would be to pop up an "add source/image" modal as a pre-requisite to a VM template with a missing source/image.
It depends on if we want to provide the ability to create a VM from a template with a "one-time" image. If we do, then we still need to provide a Create VM action fro templates with missing images.

When common templates shipped with an image, then the UI would be that this OS lists as "with image" and will have the checkbox to clone (selected by default). This is already covered in the design.

Reviewer: maybe for now we should just add a design that covers the scenario when a user tries to upload a new image for an OS that already has one.

Author: Doesn't the last image cover this? (line 55)

These OS images, and common templates in general, won’t be editable for the user.
Today, we don’t have OS images, so we need a temporary solution; offering the user a way to attach a new image that he adds to a common template, making it available on selecting that OS template field.

## Adding an OS image

![Create VM wizard - default](img/Admin-create-4.6-0-0.jpg)

On clicking "Create VM" the user lands on the wizard.
The user enters a VM name and clicks on the operating system dropdown to select an OS.

> ![OS dropdown](img/Admin-create-4.6-0-1.jpg)

Reviewer: I think we can drop the OS part of OS source available. I would think that users would make that connection

The OS dropdown would promote OSs backed by images by placing them on top.

> It will also have a search field to fetch all desired results without going through two lists.

Reviewer 2: Was the search moved to 4.7?

Author: Great catch!

This time, the user selects an OS without an available image/source.

> ![No OS image available warning](img/Admin-create-4.6-0-2.jpg)

Reviewer: The text for the action could be
```
Upload and attach an image for this operating system
```
as @glekner has pointed out - using the name of the OS can read strangely in cases where it's `centOS or higher`.

Author: Ok, @jelkosz FYI :)

Author: Should I drop the OS name for guest tools? as it is only windows.
@glekner 
Will leaving the "windows 10" help the user understand what are guest tools or is it common knowledge?

Reviewer: I think what you have for the guest tools works fine. It doesn't hurt to provide that info. But good to have @glekner confirm too

> ![No OS image available warning](img/Admin-create-4.6-0-2.jpg)

Reviewer 2: I was going to suggest maybe the warning could take place in an inline alert, since it had an action and a fair amount of text, but I think what you have here is cool.

Author: Thanks, man! I think that because it is a destructive action it is better to pop it up as a wall instead of mentioning it next to the button.

The user will be presented with a warning message notifying that no OS image/source detected.
The user can either click the "Upload and attach..." action to couple a new OS image to the OS selection, or to provide a source on the field below.

> A “source” field (previously called “source”) will present itself, indent.

Reviewer: ```suggestion
A “source” field (previously called “source”) will present itself, indented.
```
Do you mean to say boot source here?

> !["Create VM session will be discarded"](img/Admin-create-4.6-0-3.jpg)

Reviewer 2: I think the Alert convention would have this inline alert at the top of the modal, since it's there when it is opened.

Reviewer: nice catch @itsptk

Author: Not sure I follow, you mean to place this notification as an inline alert within the wizard page?

If so, I don't think I agree since:

- It is a destructive action, so it is better to block the user with a modal to confirm that to mention it next to a button where it might be missed.
- It has a lot of into going on the modal. Adding this info as an inline alert to what is an optional path in our (currently) main flow of creating a VM is simply white noise, for most of the time.

Reviewer 2: Sorry, I should have just linked to the convention initially. I was just trying to say that I think we decided if a modal opens with an inline alert already in place, it should appear at the top of the modal: http://openshift.github.io/openshift-origin-design/conventions/documentation/alerting.html#at-top-of-modal

> !["Create VM session will be discarded"](img/Admin-create-4.6-0-3.jpg)

Reviewer: What if the text for the info alert was

**Upload OS sources for cluster wide use**
By proceeding to upload you will also be able to mark your uploaded data as available across the cluster when this operating system is selected

On clicking the "Upload & Attach os source..." on the warning message link, the user will be presented with a modal.
This modal will warn the user that the "create VM" session will be discarded and all progress will be lost.
The modal will also mention the benefit of continuing, making this source available for all from now on.

## Selecting an OS with a backed OS images

> ![Making the new image available](img/Admin-create-4.6-1-0.jpg)

Reviewer: There was a bug that was filed that required a change to the info alert at the top of this page.
see here
https://xd.adobe.com/view/5a6ff231-89d3-44de-8f18-f2f4a5299636-0f3c/screen/79101d47-db7e-47e3-a0bf-43789025769c

The user browse to the image local file.
All other fields are left to default, making the image an available source when selecting "windows 10 from the OS dropdown".

![New image available](img/Admin-create-4.6-1-1.jpg)

> After the Upload and attachment completed, the source will be available for "windows 10" from now on.

Reviewer: ```suggestion
After the Upload and attach complete, the source will be available for "windows 10".
```

> ![Adding source status](img/Admin-create-4.6-1-2.jpg)

Reviewer: same comment about removing "OS" from the text here. 
Should the warning text also be in parentheses like the ones above?
Looks like there is a `O` character there too that can removed

Reviewer: This text should also read
"Source is currently uploading"

Author: "Source is currently being uploaded?"

Reviewer: "Source is currently uploading" is a bit more concise 🤷
